README.md
The __qogdata__ package is a collection of functions to manipulate [Quality of Government][qog] (QOG) data and related material. It lets the user access QOG datasets and codebooks and contains helpers to search and prepare QOG variables. The package also contains methods to merge and plot time series data, and provides a few examples of how these methods can be deployed to import World Bank and country-level indicators into QOG datasets.

[qog]: http://www.qog.pol.gu.se/

# INSTALL

Version 0.1 of the `qogdata` package is installable with the `devtools` package:

    library(devtools)
    install_github("qogdata", "briatte", dependencies = TRUE)
    require(qogdata)

Installing dependencies is particularly recommended to work with country-level data. If loading fails after running `devtools`, install dependencies manually:

    deps = c("countrycode", "foreign", "ggplot2", "maps", "WDI")
	lapply(deps, install.packages)

# HOWTO

This document describes all currently available functions, which are also internally documented. Additional notes are provided in the [repository wiki][wiki]. Please post comments, issues and suggestions in the [Issues](https://github.com/briatte/qogdata/issues) section. Feel free to [email me](mailto:f.briatte@ed.ac.uk) if you have any additional queries or ideas on how the `qogdata` package could be developed.

[wiki]: https://github.com/briatte/qogdata/wiki

## `qogdata`

1. Install and load the `xtdata` package.
2. Type `qogdata(TRUE)` to download the QOG Standard dataset.
3. Type `qogbook(TRUE)` to download the QOG Standard codebook.

__Details:__

`qogdata` simply points to a [QOG server](http://www.qogdata.pol.gu.se/) to download available versions of the QOG datasets. By default, it simply returns the path to the QOG Standard cross-section dataset:

    > qogdata()
    [1] "http://www.qogdata.pol.gu.se/data/QoG_std_cs_15May13.csv"

Set `file` to `TRUE` or to a specific filename to download the dataset:

    > QOG = qogdata(file = TRUE, format = "ts", codebook = TRUE)
    Downloading http://www.qogdata.pol.gu.se/data/QoG_std_ts_15May13.csv...
    Loaded qog_std_ts_15May13.csv (N = 14137, 1946-2012, T = 67).

Set `codebook` to `TRUE` or to a specific filename to also download the codebook with the `qogbook` function:

    > QOG = qogdata(file = TRUE, format = "ts", codebook = TRUE)
    Downloading http://www.qogdata.pol.gu.se/data/QoG_std_ts_15May13.csv...
    Loaded qog_std_ts_15May13.csv (N = 14137, 1946-2012, T = 67).
    Downloading codebook to Codebook_QoG_Std15May13.pdf...
    Codebook: Codebook_QoG_Std15May13.pdf

__Notes:__

* The QOG Standard dataset is [currently](http://www.qogdata.pol.gu.se/data/) available in CSV, SPSS and Stata formats, and other QOG datasets are available only in Stata format. `qogdata` will call `foreign` to import SPSS and Stata files, which contain variable and value labels.
* The `qogdata` package does not currently deal with the [QOG EU Regions datasets][qog-reg], which comes in Excel format and has not been updated throughout the years. However, the `xtdata` specification (explained below) is flexible enough to accommodate for regional data, so that [might happen][wiki] in the future.
* The [codebooks](http://www.qogdata.pol.gu.se/codebook/) are in PDF format. It should be easy to build an additional `qoglabel` function to call value labels extracted from Stata and stored as index files, as shown for variable labels and years of measurement in `qogfind` (below).

[qog-reg]: http://www.qog.pol.gu.se/data/datadownloads/qogeuregionaldata/

## `qogfind`

`qogfind` uses two indexes bundled with the package to help the user find variables in QOG datasets:

    > qogfind("public|administration")
    QOG Standard results:
         variable                            label ts.min ts.max ts.N ts.T cs.N cs.min cs.max
    217   gir_acs Administration and Civil Service   2004   2011  251    8   91   2006   2011
    291 irai_epru    Equity of Public Resource Use   2005   2011  532    7   80   2006   2011
    301  irai_qpa Quality of Public Administration   2005   2011  532    7   80   2006   2011
    304  irai_tac Transparency, Accountability and   2005   2011  532    7   80   2006   2011
    460  wdi_puhe Public Health Expenditure (% of    1995   2010 2960   16  187   2009   2009
    710  wvs_f115 Justifiable: avoiding a fare on    1981   2008  157   28   NA     NA     NA
    716   wvs_pet           Public self-expression   1981   2008  161   28   NA     NA     NA

The function searches through variable names and labels, as the `lookfor` command would in Stata. The `ts` columns provides years of measurement for the time-series dataset, the `cs` columns for the cross-sectional dataset. The information matches the figures reported in the _QOG Standard Codebook_ and _QOG Social Policy Codebook_. It can be easily plotted:

    qplot(data = na.omit(qogfind("ims_", version = "soc")), 
          y = label, yend = label, x = as.numeric(ts.min), xend = as.numeric(ts.max), 
          geom = "segment", size = I(6), alpha = ts.T) +
      scale_alpha("Year range") +
      theme_minimal(16) +
      labs(y = NULL, x = NULL, title = "Data availability") + 
      theme(legend.position = "bottom")

![](https://github.com/briatte/qogdata/raw/master/example1.png)

## `qogjoin`

`qogjoin` is a highly idiosyncratic method to join historical to recent states in the QOG Standard time series dataset. Only states that went through a single territorial departure (e.g. France/Algeria or Pakistan/Bangladesh) are joined: more complex state partition cases like North/South Sudan and North/South Yemen are left aside.

In the current Standard version of the dataset (15 May 2013), this will cause a merge between the split versions of Ethiopia, France, Malaysia and Pakistan. This will make the dataset more backward compatible with older versions of the data where this separation did not exist.

## `xtdata`, `xtset`, `xt`, `xtdes`

`xtdata` and its related functions are a way to specify panel data properties into the `xtdata` attribute of a data frame attribute, which makes it possible to:

* safely merge several panel datasets with `xtmerge`, a merge function that first checks whether the unit type and time period formats match, and looks for matches if they do not.
* pass some default information to plot methods for panel data, which is minimally illustrated in `xtmap` and forthcoming as a larger series of `xtplot` methods for `ggplot2`.

Please consult the [repository wiki][wiki] for details on `xt` methods in the `qogdata` package, and how it might turn into a proper class by version 1.0. The method is intended for use with panel data: users who want to work with full-fledged time series should use the `zoo` or `xts` packages.

## `xtshift`, `xtlag`, `xtlead`, `xtdecay`, `xttse`

`xtshift`, `xtlag` and `xtlead` are [functions to shift (lag or lead) a panel variable](http://christophergandrud.blogspot.fr/2013/05/slide-one-function-for-laglead.html). `xtdecay` and `xttse` (time since event) are additional [time series functions for panel data](http://www.zmjones.com/panel-ts.html). These functions, as well as several other little utilities used in the `qogdata` package, are based on code found online (see the package documentation for the sources).

## `xtmerge`, `xtsubset`, `xtsample`

`xtmerge` performs a merge of two panel datasets based on their `xtdata` attributes, checking for identically formatted data identifiers and time periods before performing the merge, which otherwise works like [`merge`][merge] in base R.

[merge]: http://www.rdocumentation.org/packages/base/functions/merge

When `xtmerge` is provided datasets of type `country` with different country code formats, it runs the `xtcountry` helper function to determine the best ISO-3N conversion match and performs the merge on the new `iso3n` variable.

`xtsubset` is a wrapper for `subset` that preserves the `xtdata` attribute of a data frame. `xtsample` is a wrapper for `sample` that selects all observations from a random sample of unique identifiers (thereby preserving the time structure of the data).

## `xtmap`, `xtplot`

`xtmap` calls the `countrycode`, `maps` and `ggplot2` packages to draw choropleth maps, with helper transformations for time series. The examples provided below use QOG cross-sectional data:

    QOG = qogdata(tempfile(), version = "bas", variables = c("ccodealp", "undp_hdi", "ihme_nm"))

The result is a `ggplot2` object with the [map and data plotted together][map_data]:

[map_data]: http://www.rdocumentation.org/packages/ggplot2/functions/map_data

    xtmap(subset(QOG, ccodealp != "RUS"), "ihme_nm", continent = "Asia", iso3n = "ccode") +
      ggtitle("Neonatal Mortality Rate per 1,000 births (IHME, 2009))")

![](https://github.com/briatte/qogdata/raw/master/example2.png)

Use the `quantize` option to map quantiles of a continuous variable on the fly:

    xtmap(QOG, "undp_hdi", quantize = 3, continents = c("Africa", "Asia"), iso3n = "ccode") +
      scale_fill_brewer("", palette = "RdYlBu", labels = c("Low", "Med", "High")) +
      ggtitle("Human Development Index (UNDP, 2009-2010)")

![](https://github.com/briatte/qogdata/raw/master/example3.png)

The function matches countries to geographic information from the `world` map provided in the `maps` package, and currently suffers from a little [map projection bug][bug] as soon as you include Russia. It also adds continents and regions with the `countrycode` package to allow plots of specific geographic areas.

[bug]: https://github.com/briatte/qogdata/issues/1

When provided with a data frame carrying the `xtdata` attribute, the function currently uses the maximum time period as `t`, as with 'most recent year' with `country`-level data. If `quantize.t` is used to create time intervals, the function plots facets of maps.

`xtplot` is a stub for a similar function that plots time series out of data frames carrying an `xtdata` attribute.

## `xtmissing`

`xtmissing` plots a basic tile plot of missing and nonmissing values in a data frame with an `xtdata` attribute, with observations in rows and ordered time periods in columns.

## `get_eurostat`

`get_eurostat` is a soon-to-be-added downloader for Eurostat data. See the [repository wiki][wiki] for development notes, or see the [`eurostat_r`][eurostat_r] script to bulk import Eurostat data in R.

[eurostat_r]: https://github.com/toprach/eurostat_r

## `get_wdi`

`get_wdi` calls the [WDI](http://cran.r-project.org/web/packages/WDI/) package to download one or more [World Development Indicators](http://data.worldbank.org/data-catalog/world-development-indicators) provided through the World Bank API. The result is a data frame with an `xdtata` attribute ready for merging onto ISO-3N country codes, converted from the ISO-3C country codes returned by `WDI`.

This function makes it easy to update a WDI within a QOG dataset to the latest measurements, and to compare measurement differences between the latest QOG and WDI series:

![](https://github.com/briatte/qogdata/raw/master/example4.png)

The WDI data for this plot was retrieved with `get_wdi` as follows:

    WDI = get_wdi(x = "SH.XPD.PCAP.PP.KD", add = "income")

The code for the comparative measurements plot is in the documentation of the `get_wdi` function.

## `get_uds`

`get_uds` downloads the [Unified Democracy Scores](http://www.unified-democracy-scores.org/) (UDS) from Pemstein _et al._. The result is a data frame with an `xdtata` attribute ready for merging onto Correlates of War (COW) numeric country codes.

## `get_gwpt`

`get_gwpt` downloads [state independence][statei] data from Gleditsch & Ward and  [state coups][statec] data from Powell & Thyne. The result is a data frame with an `xdtata` attribute that is idiosyncratic (Gleditsch and Ward country codes), but might still be merged with `xtcountry` and `xtmerge` (see above).

[statei]: http://privatewww.essex.ac.uk/~ksg/statelist.html
[statec]: http://www.uky.edu/~clthyn2/coup_data/home.htm

# CREDITS

`qogdata` takes inspiration from two [QOG packages for Stata users](http://www.qog.pol.gu.se/data/dataextras/forstatausers/), [`qog`](http://ideas.repec.org/c/boc/bocode/s457283.html) by Christoph Thewes and [`qogbook`](http://ideas.repec.org/c/boc/bocode/s457599.html) by Richard Svensson. Further credits are due to the [authors of the QOG datasets][qog-team], and a list of related R packages appears on the [wiki][wiki].

[qog-team]: http://www.qog.pol.gu.se/aboutus/Organization/


NEWS
Version 0.1.5 (2013-08-10)
--------------------------

minor fixes and additions

xt functions:

* xtdecay: linear decay
* xttse: time since event
* xtshift, xtlag, xtlead: lag/lead variables
* xtdes, xtsubset, xtsample methods
* xtcountry, xtcheck internals
* xtplot stub

utils:

* quantize, std01
* try_require, sort_df (by Hadley Wickham)
* panel.tse (by Zachary M. Jones)
* decay (by Zachary M. Jones)
* shift (by TszKin Julian Chan)

Version 0.1 (2013-08-05)
------------------------

initial release

qog functions: 

* qogdata: download datasets
* qogbook: download codebooks
* qogfind: search variables
* qogjoin: merge historical states

xt functions:

* xtdata specification
* xt, xtset internals
* xtmerge method
* xtmissing values plot
* xtmap

get functions:

* get_wdi (World Bank indicators, using WDI package)
* get_uds (Pemstein et al. democracy scores)
* get_gwpt (Gleditsch and Ward/Powell and Thyne state variables)


R/qog.R
#' Get Quality of Government datasets in \code{xtdata} format
#'
#' Function to download Quality of Government (QOG) data and load it as a data 
#' frame in R. The result carries an \code{\link{xtdata}} attribute that can be 
#' passed to the \code{\link{xtmerge}} panel data method. Please visit the 
#' QOG Institute website at \url{http://www.qog.pol.gu.se/} for a presentation 
#' of QOG research.
#'
#' @export
#' @aliases get_qog
#' @param file a filename to save the dataset at. 
#' If set to \code{FALSE} (the default), the function just returns the link 
#' to the dataset. 
#' If set to \code{TRUE}, the server filename of the dataset is used, which 
#' returns either a CSV file if \code{version} is set to \code{std}, or
#' a Stata \code{dta} file otherwise. See 'Details'.
#' @param replace whether to overwrite the dataset even if a file already 
#' exists at the download location. Defaults to \code{FALSE}.
#' @param path a folder path to prepend to the filename and to the codebook
#' if \code{codebook} is not \code{FALSE}.
#' @param version the QOG version: 
#' \code{std} (Standard), \code{soc} (Social Policy), \code{bas} (Basic) 
#' or \code{exp} (Expert). Defaults to \code{std}. See 'Details'.
#' @param format the QOG format, usually \code{cs} for cross-sectional data
#' or \code{ts} for time series in the \code{std} and \code{bas} versions. 
#' See 'Details' for the full list of specifications. Defaults to \code{cs}.
#' @param codebook whether to download the codebook. Calls \code{qogbook} by 
#' passing the \code{codebook}, \code{version} and \code{path} arguments to it, 
#' where \code{codebook} is treated as the filename for the codebook. 
#' Defaults to \code{FALSE}.
#' @param variables a selection of variables to import. \code{ccode} ISO-3N 
#' country codes \code{ccode} and \code{year} identifiers will be forced into
#' the output if relevant.
#' @param years a selection of years to import. Effective only with
#' the \code{ts}, \code{tsl} or \code{ind} formats.
#' @param ... other arguments supplied to the import method, which is 
#' \code{read.csv} by default, 
#' or \code{\link[foreign]{read.dta}} if \code{file} is a Stata \code{dta} dataset,
#' or \code{\link[foreign]{read.spss}} if \code{file} is a SPSS \code{sav} dataset.
#' @details This version of the package handles all four QOG datasets:
#' \tabular{lcl}{
#'  QOG Standard \tab \code{std} \tab 15 May 2013\cr
#'  QOG Social Policy \tab \code{soc} \tab 4 April 2012\cr
#'  QOG Basic \tab \code{bas}): \tab 28 March 2011\cr
#'  QOG Expert Survey \tab \code{exp} \tab 3-6 September 2012\cr
#'  URL: \tab \tab \url{http://www.qog.pol.gu.se}\cr
#' }
#'
#' Each QOG dataset is available in a variety of data formats:
#' 
#' \itemize{
#'   \item QOG datasets \code{std} and \code{bas} 
#'   require format \code{cs} (cross-section) 
#'   or \code{ts} (time series).
#'   \item QOG dataset \code{soc} 
#'   requires format \code{cs}, \code{tsl} (time series, long) 
#'   or \code{tsw} (time series, wide)
#'   \item QOG dataset \code{exp} 
#'   requires format \code{cntry} (country-level) 
#'   or \code{ind} (individual survey)
#' }
#'
#' The QOG Standard series comes in CSV, SPSS and Stata file formats, CVS being
#' the only format that contains numeric codes instead of QOG value labels. 
#' Datasets outside of the QOG Standard series are available only as Stata items 
#' and require that \code{file} ends in \code{.dta} when \code{version} is not 
#' \code{std}. The only exception is dataset \code{csyom}, which automatically
#' sets \code{version} to \code{std} and requires \code{file} to end
#' in \code{.csv}. Filenames with inadequate extensions will be modified to 
#' conform to these expectations if they do not.
#' @seealso \code{\link{qogbook}}, \code{\link[foreign]{read.dta}}, \code{\link[foreign]{read.spss}}
#' @author Francois Briatte \email{f.briatte@@ed.ac.uk}
#' @examples
#' # Show URL to QOG Standard cross-section.
#' qogdata()
#' # Show URL to QOG Social Policy time series, long format.
#' qogdata(version = "soc", format = "tsl")
#' ## Download codebook and recent years from QOG Basic cross-section (not run).
#' # QOG = qogdata(file = "qog.cs.txt", version = "bas", format = "cs", 
#' #         years = 2002:2012, codebook = TRUE)
#' ## Download QOG Standard cross-section years of measurement (not run).
#' # QOG = qogdata(tempfile(fileext = ".csv"), format = "csyom")
#' ## Show QOG years of measurement for Gini coefficient (not run).
#' # table(QOG$wdi_gini)
#' @keywords qog

qogdata <- function(file = FALSE, replace = FALSE, codebook = FALSE, path = "",
                version = "std", format = "cs", 
                variables = NULL, years = NULL, ...) {
  try_require("foreign")
  #
  # currently available
  #
  versions = list(
    std = c("ts", "cs", "csyom"),
    bas = c("ts", "cs"),
    exp = c("ctry", "ind"),
    soc = c("cs", "tsl", "tsw"))
  #
  # correct version
  #
  if(!version %in% names(versions)) {
    stop("Invalid version: use one of ", 
         paste0(names(versions), collapse = ", "))
  }
  #
  # correct format
  #
  if(format == "csyom") {
    version = "std"
    if(!grepl(".csv$|.txt$", file))
      file = gsub("(\\.|\\w){4}$", ".csv", file)
  }
  if(!format %in% unlist(versions[version])) {
    stop("Invalid format: use one of ", 
         paste0(unlist(versions[version]), collapse = ", "))
  }
  #
  # automatic filename
  #
  if(isTRUE(file)) {
    file = paste0("qog_", 
                  version, 
                  "_", 
                  format, 
                  ifelse(version == "std", 
                         paste0("_", "15May13.csv"), 
                         ".dta")
                  )
  }
  else {
    if(is.character(file) & version != "std" & !grepl(".dta$", file)) {
      file = gsub("(\\.|\\w){4}$", ".dta", file)
      warning("QOG datasets other than std are available only as Stata files.\n",
           "  The filename that you specified was modified to ", file)      
    }
  }
  if(is.character(path))
    if(nchar(path) > 0) file = paste(path, file, sep = "/")
  #
  # online source
  #
  link = paste0("http://www.qogdata.pol.gu.se/data/",
                ifelse(version == "std", "QoG", "qog"),
                "_", version, "_", format, 
                ifelse(version == "std", paste0("_", "15May13"), ""),
                ifelse(version == "std" & grepl("csv|dta|sav", file), 
                       substring(file, nchar(file) - 3),
                       ".dta")
  )
  if(is.logical(file)) {
    return(link)
  }
  else {
    if(replace || !file.exists(file)) {
      message("Downloading ", link, "...")
      download.file(link, file, mode = "wb", quiet = TRUE)
    }
    else {
      message("Loading from disk...")
    }
  }
  #
  # reader call
  #
  read = "read.csv"
  args = list(file = file, ...)
  # foreign or not
  if(!grepl(".dta$|.sav$", file)) 
    args["sep"] = ";"
  # stata args
  if(grepl(".dta$", file)) {
    read = "read.dta"
    if(is.null(unlist(args["warn.missing.labels"])))
      args["warn.missing.labels"] = FALSE
  }
  # spss args
  if(grepl(".sav$", file)) {
    read = "read.spss"
    if(is.null(unlist(args["to.data.frame"])))
      args["to.data.frame"] = TRUE
  }
  data = do.call(read, args)
  #
  # selected variables
  #
  uids = c("ccode", "ccodealp", "cname", "ccodecow", "ccodewb")
  pids = uids %in% names(data)

  # avoid ts bug
  if(grepl("ts|tsl", format) & !any(pids))
    stop("You are trying to load a QOG dataset as time series, but it has no identifier variable.")

  # avoid ts bug
  if(grepl("ts|tsl", format) & !"year" %in% names(data))
    stop("You are trying to load a QOG dataset as time series, but it has no year variable.")
 
  if(!is.null(variables)) {
    if(grepl("ts|tsl", format) & !"year" %in% variables) {
      warning("Forcing year identifier into the dataset.")
      variables = c("year", variables)
    }
    if(grepl("std|bas|soc", version) & !"ccode" %in% variables) {
      warning("Forcing ccode identifier into the dataset.")
      variables = c("ccode", variables)
    }
    data = data[, names(data) %in% variables]
  }
  #
  # selected years
  #
  if(!is.null(years) && format %in% c("ts", "tsl", "ind"))
    data = data[data$year %in% years, ]
  #
  # message
  #
  message("Loaded ", file, " (N = ", nrow(data),
          ifelse(format %in% c("ts", "tsl", "ind"),
                 paste0(", ", min(data$year), 
                        "-", max(data$year), 
                        ", T = ", length(unique(data$year))),
                 ""),
          ").")
  #
  # grab codebook
  #
  if(isTRUE(codebook) || grepl(".pdf", codebook))
    qogbook(codebook, version, path, replace)
  #
  # xtdata spec
  #
  pids = uids[pids] 
  if(format == "ts" | format == "tsl") {
    data = xtset(data, 
                 data = c(pids[1], "year", pids[-1]), 
                 spec = c("iso3n", "year"), 
                 type = "country", 
                 name = "Quality of Government, time series data"
    )
  }
  #
  # finish line
  #
  return(data)
}

#' Download Quality of Government codebooks
#'
#' Function to download Quality of Government (QOG) codebooks. Please visit 
#' the QOG Institute website at \url{http://www.qog.pol.gu.se/} for a 
#' presentation of QOG research.
#'
#' @export
#' @param file a filename to save the codebook at. 
#' If set to \code{TRUE}, the name of the codebook on the QOG server will be used. 
#' If set to \code{FALSE} (the default), the function only returns the link to 
#' the dataset. The filename must end in \code{.pdf}.
#' @param replace whether to download the dataset even if a file already exists 
#' at the download location. Defaults to \code{FALSE}.
#' @param path a folder path to append to the filename.
#' @param version the QOG version: \code{std} (Standard), 
#' \code{soc} (Social Policy), \code{bas} (Basic) or \code{exp} (Expert). 
#' Defaults to \code{std}.
#' @details The function mimics Richard Svensson's \code{qogbook} Stata command.
#' @seealso \code{\link{qogdata}}
#' @author Francois Briatte \email{f.briatte@@ed.ac.uk}
#' @examples
#' # Show the URL to the QOG Standard dataset codebook.
#' qogbook()
#' ## Download QOG Standard codebook with default filename (not run).
#' # qogbook(file = TRUE)
#' ## Download QOG Basic dataset codebook to specific filename (not run).
#' # qogbook(file = "qog.basic.codebook.pdf", version = "bas")
#' @keywords qog

qogbook <- function(file = FALSE, version = "std", path = "", replace = FALSE) {
  if(!version %in% c("std", "bas"))
    stop("Codebook available only for versions bas, std")
  
  link = "http://www.qogdata.pol.gu.se/data/Codebook_QoG_Std15May13.pdf"
  if(version == "bas") {
    link = "http://www.qogdata.pol.gu.se/codebook/codebook_basic_20120608.pdf"
  }
  if(isTRUE(file) & version == "bas") {
    file = "codebook_basic_20120608.pdf"
  }
  else if(isTRUE(file)) {
    file = "Codebook_QoG_Std15May13.pdf"
  }
  #
  # path
  #
  if(is.character(path))
    if(nchar(path) > 0) file = paste(path, file, sep = "/")
  #
  # download
  #
  if(is.logical(file)) {
    return(link)
  }
  else if(!grepl(".pdf$", file)) {
    stop("Please specify a .pdf codebook filename or set file = TRUE.")
  }
  else {  
    if(replace || !file.exists(file)) {
      message("Downloading codebook to ", file, "...")
      download.file(link, file, mode = "wb", quiet = TRUE)
    }
    message("Codebook: ", file)
  }
}

#' Find Quality of Government variables
#'
#' Function to perform a \code{regex} search on QOG variable names and labels. 
#' A few labels are missing for strictly cross-sectional variables.
#'
#' @export
#' @param ... keywords or \code{regex} phrases passed to \code{grepl}.
#' @param version the QOG version to search: either \code{std} (the default) or \code{soc}.
#' @param compact whether to limit the labels returned to 32 characters. Defaults to \code{TRUE} for better console output.
#' @param show which variables to show for years of measurement: \code{cs} (cross-sectional), \code{ts} (time series), or \code{all} (the default).
#' @return a data frame containg matching variables, described by their names, labels and years of measurement in the time series (\code{ts}) cross-sectional (\code{cs}) datasets. The information should match the ranges indicated in the \emph{QOG Standard Codebook} and \emph{QOG Social Policy Codebook}.
#' @references
#' Svensson, Richard, Stefan Dahlberg, Staffan Kumlin & Bo Rothstein. 
#' 2012. \emph{The QoG Social Policy Dataset}, version 4Apr12. 
#' University of Gothenburg: The Quality of Government Institute, 
#' \url{http://www.qog.pol.gu.se}.
#' 
#' Teorell, Jan, Nicholas Charron, Stefan Dahlberg, Soren Holmberg, 
#' Bo Rothstein, Petrus Sundin & Richard Svensson. 2013. 
#' \emph{The Quality of Government Dataset}, version 15May13. 
#' University of Gothenburg: The Quality of Government Institute, 
#' \url{http://www.qog.pol.gu.se}.
#' @seealso \code{\link{grep}}, \code{\link{qogdata}}
#' @author Francois Briatte \email{f.briatte@@ed.ac.uk}
#' @examples
#' # QOG Standard search.
#' qogfind("regime", "institutions")
#' # QOG Standard search, with regex syntax.
#' qogfind("public|administration")
#' # QOG Social Policy search, showing cross-sectional information only.
#' head(qogfind("^socx", version = "soc", show = "cs", compact = FALSE))
#' # QOG Standard variables featured only in the cross-sectional version.
#' qogfind("*")[is.na(qogfind("*")$ts.N), ]
#' @keywords qog

qogfind <- function(..., version = "std", compact = TRUE, show = "all") {
  x = paste0(c(...), collapse = "|")
  data(qog.index)
  if (version == "std") {
    message("QOG Standard results")
    r = qog.std.index[grepl(x, qog.std.index$variable, ignore.case = TRUE) | 
                        grepl(x, qog.std.index$label, ignore.case = TRUE), ]
  }
  if (version == "soc") {
    message("QOG Social Policy results")
    r = qog.soc.index[grepl(x, qog.soc.index$variable, ignore.case = TRUE) | 
                        grepl(x, qog.soc.index$label, ignore.case = TRUE), ]
  }
  if (version == "bas") {
    message("QOG Basic results")
    r = qog.bas.index[grepl(x, qog.bas.index$variable, ignore.case = TRUE) | 
                        grepl(x, qog.bas.index$label, ignore.case = TRUE), ]
  }
  r$variable = as.character(r$variable)
  if(compact) r$label = substr(r$label, 1, 32)
  if(show == "cs") r = r[, !grepl("ts.", names(r))]
  if(show == "ts") r = r[, !grepl("cs.", names(r))]
  return(r)
}

#' Join historical and recent states in QOG Standard time series data
#'
#' Function to plot maps of Quality of Government (QOG) data. Requires the \code{ggplot2} and \code{maps} packages.
#'
#' @export
#' @param data a QOG Standard time series dataset, or any data frame with \code{cname} (country) and \code{year} information coded as in the QOG Standard time series dataset.
#' @param country the country name to join data over. Requires the \code{cname} variable.
#' @details The function will try to find two series of country-year observations that both match the \code{country} argument. Within the QOG Standard time series dataset, this will match historical states like "France (-1962)" to modern states like "France (1963-)". The function will then create a new variable out of both series, joined at their separation years, and set its country code attributes to the most recent ones. See Appendix A of the \emph{QOG Standard Codebook} for details on historical states in the QOG Standard time series dataset.
#' @return a data frame with country-year observations
#' @author Francois Briatte \email{f.briatte@@ed.ac.uk}
#' @references Teorell, Jan, Nicholas Charron, Stefan Dahlberg, Soren Holmberg, 
#' Bo Rothstein, Petrus Sundin & Richard Svensson. 2013. 
#' \emph{The Quality of Government Dataset}, version 15May13. 
#' University of Gothenburg: The Quality of Government Institute, 
#' \url{http://www.qog.pol.gu.se}.
#' @examples
#' # Load QOG demo datasets.
#' data(qog.demo)
#' QOG = qog.ts.demo
#' QOG = qogjoin(QOG, "Ethiopia")
#' QOG = qogjoin(QOG, "France")
#' QOG = qogjoin(QOG, "Malaysia")
#' QOG = qogjoin(QOG, "Pakistan")
#' @keywords qog

qogjoin <- function(data, country = NULL) {
  stopifnot("cname" %in% names(data))
  stopifnot("year" %in% names(data))
  stopifnot(is.character(country))
  x = data$cname[grepl(country, data$cname)]
  x = unique(x)
  if(length(x) == 0)
    stop("No country match for ", country)
  else if(length(x) == 1)
    stop("Single country match for ", country)
  else if(length(x) > 2)
    stop("More than two country matches: ", paste0(x, collapse = ", "))
  
  t = xt(data)
  
  y = as.numeric(gsub("\\D", "", x))
  min = x[order(y) == 1]
  max = x[order(y) == 2]
  message("Joining ", min, " to ", max)
  
  one = data[data$cname == min & data$year <= y[order(y) == 1], ]
  two = data[data$cname == max & data$year >= y[order(y) == 2], ]
  new = rbind(one, two)
  
  new$cname = country
  
  # country codes
  if("ccode" %in% names(data))
    new$ccode = unique(data$ccode[data$cname == max])
  if("ccodealp" %in% names(data))
    new$ccodealp = unique(data$ccodealp[data$cname == max])
  if("cname_year" %in% names(data))
    new$cname_year = paste(new$cname, new$year)
  if("ccodealp_year" %in% names(data))
    new$ccodealp_year = paste(new$ccodealp, new$year)
  if("ccodewb" %in% names(data))
    new$ccodewb = unique(data$ccodewb[data$cname == max])
  
  message(country, " now runs over ", nrow(new), " country-year observations, ",
          paste0(range(new$year), collapse = "-"))

  data[, "cname"] = as.character(data[, "cname"])
  data[data$cname == max, ] = new
  data = subset(data, cname != min)
  data[, "cname"] = factor(data[, "cname"])
  
  data = xtset(data,
               data = t$data,
               spec = t$spec,
               name = t$name,
               url = t$url,
               quiet = FALSE)
  return(data)
}
